src/features/contracts/contracts-directory.ts

```diff
@@ -7,14 +7,79 @@ import * as util from 'util';
 import * as stream from 'stream';
 import * as path from 'path';
 import * as os from 'os';
+import validator from 'validator';
 import type { RepositoryInfo, Contract } from './index';
+import { getBase64DataUri } from '../../lib/utils';
+import { captureException } from '../../infra/error-handling';
 
 const pipeline = util.promisify(stream.pipeline);
 const exists = util.promisify(fs.exists);
 const mkdir = util.promisify(fs.mkdir);
 
 const CONTRACTS_BASE_DIR = path.join(os.tmpdir(), 'contracts');
 
+// All assets that are stored together with the contract are encoded and stored in a dataurl format.
+const handleLocalAssetUrl = async (assetUrl: string): Promise<string> => {
+	switch (path.extname(assetUrl)) {
+		case '.svg': {
+			return getBase64DataUri(assetUrl, 'image/svg+xml');
+		}
+		case '.png': {
+			return getBase64DataUri(assetUrl, 'image/png');
+		}
+		default:
+			return '';
+	}
+};
+
+const normalizeAssets = async (
+	contractFilepath: string,
+	assets: Contract['assets'],
+) => {
+	if (!assets || _.isEmpty(assets)) {
+		return assets;
+	}
+
+	const normalizedAssets: Contract['assets'] = {};
+
+	await Promise.all(
+		Object.entries(assets).map(async ([key, asset]) => {
+			if (validator.isURL(asset.url)) {
+				normalizedAssets[key] = asset;
+				return;
+			}
+
+			// Convert from relative to absolute path for the asset file and make sure it doesn't try to access files outside of the contract folder.
+			const contractDir = path.dirname(contractFilepath);
+			const normalizedUrl = path.resolve(contractDir, asset.url);
+			if (
+				!normalizedUrl.startsWith(contractDir) ||
+				(await (await fs.promises.lstat(normalizedUrl)).isSymbolicLink())
```

Comment:
Does this check if the path included a symlink or only if the final entry is a symlink? Because it's the former we'd need to do this properly (the latter would just give a false sense of safety since we could already have escaped the contract dir). What might be a better approach is to resolve the absolute path and then compare that to the contracts dir, that way you can have symlinks that are internal to the contract dir but not ones that leave it (well, technically they can leave it if they come back in.. but I guess that's fine if you wanna do that? :S  )

Reply:
@Page- updated, let me know what you think. I'll give it a go in bob now as well, but the tests are happy